native/core/src/execution/operators/parquet_writer.rs

```diff
@@ -49,6 +53,134 @@ use parquet::{
 
 use crate::execution::shuffle::CompressionCodec;
 
+/// Enum representing different types of Arrow writers based on storage backend
+enum ParquetWriter {
+    /// Writer for local file system
+    LocalFile(ArrowWriter<File>),
+    /// Writer for HDFS or other remote storage (writes to in-memory buffer)
+    /// Contains the arrow writer, HDFS operator, and destination path
+    /// an Arrow writer writes to in-memory buffer the data converted to Parquet format
+    /// The opendal::Writer is created lazily on first write
+    Remote(
+        ArrowWriter<Cursor<Vec<u8>>>,
+        Option<opendal::Writer>,
+        Operator,
+        String,
+    ),
+}
+
+impl ParquetWriter {
+    /// Write a RecordBatch to the underlying writer
+    async fn write(
+        &mut self,
+        batch: &RecordBatch,
+    ) -> std::result::Result<(), parquet::errors::ParquetError> {
+        match self {
+            ParquetWriter::LocalFile(writer) => writer.write(batch),
+            ParquetWriter::Remote(
+                arrow_parquet_buffer_writer,
+                hdfs_writer_opt,
+                op,
+                output_path,
+            ) => {
+                // Write batch to in-memory buffer
+                arrow_parquet_buffer_writer.write(batch)?;
+
+                // Flush and get the current buffer content
+                arrow_parquet_buffer_writer.flush()?;
+                let cursor = arrow_parquet_buffer_writer.inner_mut();
+                let current_data = cursor.get_ref().clone();
+
+                // Create HDFS writer lazily on first write
+                if hdfs_writer_opt.is_none() {
+                    let writer = op.writer(output_path.as_str()).await.map_err(|e| {
```

Comment:
Should we use a generic ObjectStore for writing instead of directly using opendal Operator?